File: web/network.go
/*
 * Copyright (C) 2020 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package web

import (
	"encoding/json"
	"fmt"
	"github.com/CanonicalLtd/device-config/service/network"
	"io"
	"net/http"
	"strings"
)

// InterfaceConfig defines the configuration of an interface
type InterfaceConfig struct {
	Use         bool     `json:"use"`
	Method      string   `json:"method"`
	Interface   string   `json:"interface"`
	NameServers []string `json:"nameServers"`
	Address     string   `json:"address"`
	Mask        string   `json:"mask"`
	Gateway     string   `json:"gateway"`
}

// Network is the API to get the network interface config
func (srv Web) Network(w http.ResponseWriter, r *http.Request) {
	// Get the current netplan settings
	netYAML := srv.NetSrv.Current()

	// Get the hardware interfaces
	hardware, err := network.Interfaces()
	if err != nil {
		formatStandardResponse("interfaces", err.Error(), w)
		return
	}

	// Decode the configuration for each hardware interface
	interfaces := []InterfaceConfig{}
	for _, iface := range hardware {
		cfg := InterfaceConfig{Interface: iface.Name, Use: false}

		// Get the current interface config
		eth, ok := netYAML.Network.Ethernets[iface.Name]
		if !ok {
			// The interface is not configured
			interfaces = append(interfaces, cfg)
			continue
		}

		srv.decodeNetplanInterface(&cfg, eth)

		interfaces = append(interfaces, cfg)
	}

	// Create the JSON response
	formatNetworkResponse(interfaces, w)
}

// NetworkApply is the API to apply the current network configuration
func (srv Web) NetworkApply(w http.ResponseWriter, r *http.Request) {
	// Store the interface config
	if err := srv.NetSrv.Apply(); err != nil {
		formatStandardResponse("interface-apply", err.Error(), w)
		return
	}
	formatStandardResponse("", "", w)
}

func (srv Web) decodeNetplanInterface(cfg *InterfaceConfig, eth network.Ethernet) {
	// Parse the config
	cfg.Use = true
	cfg.Gateway = eth.Gateway4
	if len(eth.DHCP4) > 0 {
		cfg.Method = "dhcp"
	} else {
		cfg.Method = "manual"
	}
	if eth.NameServers != nil {
		cfg.NameServers = eth.NameServers["addresses"]
	}
	if eth.Addresses != nil {
		addressPlusMask := strings.Split(eth.Addresses[0], "/")
		cfg.Address = addressPlusMask[0]
		if len(addressPlusMask) > 1 {
			cfg.Mask = addressPlusMask[1]
		}
	}
}

// NetworkInterface is the API to store the network interface configuration
func (srv Web) NetworkInterface(w http.ResponseWriter, r *http.Request) {
	req := srv.decodeNetworkInterface(w, r)
	if req == nil {
		return
	}

	eth := srv.encodeNetplanInterface(req)

	// Store the interface config
	if err := srv.NetSrv.Store(eth); err != nil {
		formatStandardResponse("interface-store", err.Error(), w)
		return
	}
	formatStandardResponse("", "", w)
}

func (srv Web) encodeNetplanInterface(req *InterfaceConfig) network.Ethernet {
	// Encode the interface format into the netplan format
	eth := network.Ethernet{}
	eth.Use = req.Use
	eth.Name = req.Interface
	if req.Method == "dhcp" {
		eth.DHCP4 = "true"
	} else {
		eth.DHCP4 = ""
		eth.NameServers = map[string][]string{"addresses": req.NameServers}
		eth.Gateway4 = req.Gateway

		addr := req.Address
		if len(req.Mask) > 0 {
			addr = fmt.Sprintf("%s/%s", req.Address, req.Mask)
		}
		eth.Addresses = []string{addr}
	}
	return eth
}

func (srv Web) decodeNetworkInterface(w http.ResponseWriter, r *http.Request) *InterfaceConfig {
	// Decode the JSON body
	req := InterfaceConfig{}
	err := json.NewDecoder(r.Body).Decode(&req)
	switch {
	// Check we have some data
	case err == io.EOF:
		formatStandardResponse("data", "No interface data supplied.", w)
		return nil
		// Check for parsing errors
	case err != nil:
		formatStandardResponse("decode-json", err.Error(), w)
		return nil
	}
	return &req
}

File: webapp/src/components/Network.js
/*
 * Copyright (C) 2020 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, { Component } from 'react';
import {formatError, T} from './Utils';
import AlertBox from './AlertBox';
import api from "./api"

const methods = ['dhcp', 'manual']

class Network extends Component {
    constructor(props) {
        super(props)
        this.state = {
            selected: null,
            interfaces: [],
            interface: {},
            error: '',
            message: '',
        };
    }

    componentDidMount() {
        this.getNetworkConfig()
    }

    getNetworkConfig = () => {
        api.networkGet().then(response => {
            // Skip the hidden interfaces
            let interfaces = response.data.interfaces.filter(iface => {
                return !this.props.config.hideInterfaces.includes(iface.interface)
            })

            this.setState({interfaces: interfaces, error: ''})

            if ((!this.state.selected) && (response.data.interfaces.length > 0)) {
                this.getInterface(interfaces[0].interface)
                this.setState({selected: interfaces[0].interface})
            }
        })
        .catch(e => {
            this.setState({error: formatError(e.response.data)});
        })
    }

    handleSelectTab = (e) => {
        e.preventDefault()
        let iface = e.target.getAttribute('data-key')
        this.getInterface(iface)
        this.setState({selected: iface, error: '', message:''})
    }

    handleUseChange = (e) => {
        e.preventDefault()
        let iface = this.state.interface
        iface.use = !iface.use
        this.setState({interface: iface})
    }

    handleMethodChange = (e) => {
        e.preventDefault()
        let iface = this.state.interface
        iface.method = e.target.value
        this.setState({interface: iface})
    }

    handleGatewayChange = (e) => {
        e.preventDefault()
        let iface = this.state.interface
        iface.gateway = e.target.value
        this.setState({interface: iface})
    }

    handleAddressChange = (e) => {
        e.preventDefault()
        let iface = this.state.interface
        iface.address = e.target.value
        this.setState({interface: iface})
    }

    handleMaskChange = (e) => {
        e.preventDefault()
        let iface = this.state.interface
        iface.mask = e.target.value
        this.setState({interface: iface})
    }

    handleDNSChange = (e) => {
        e.preventDefault()
        let iface = this.state.interface
        iface.nameServers = e.target.value.split(',')
        this.setState({interface: iface})
    }

    handleSave = (e) => {
        e.preventDefault()

        // Save the interface config
        api.networkUpdate(this.state.interface).then(response => {
            this.getNetworkConfig()
            this.setState({message: T('interface-updated'), error: ''})
        })
        .catch(e => {
            this.setState({error: formatError(e.response.data), message: ''});
        })
    }

    handleApply = (e) => {
        e.preventDefault()

        // Apply the config
        api.networkApply().then(response => {
            this.getNetworkConfig()
            this.setState({message: T('interface-updated'), error: ''})
        })
        .catch(e => {
            this.setState({error: formatError(e.response.data), message: ''});
        })
    }

    getInterface(iface) {
        let matches = this.state.interfaces.filter(i => {
            return i.interface===iface
        })
        if (matches.length>0) {
            this.setState({interface: matches[0]})
        }
    }

    renderUse() {
        if (this.state.interface.use) {
            return (
                <div>
                    <a href="#use" className="p-button--base has-icon" onClick={this.handleUseChange}><img src="/static/images/checkbox_checked_16.png" alt="checked" /></a>
                    <span>{T('use')}</span>
                </div>
            )
        } else {
            return (
                <div>
                    <a href="#use" className="p-button--base has-icon" onClick={this.handleUseChange}><img src="/static/images/checkbox_unchecked_16.png" alt="unchecked"  /></a>
                    <span>{T('use')}</span>
                </div>
            )
        }
    }

    renderMessage() {
        if (this.state.message) {
            return <AlertBox type="positive" message={this.state.message}/>
        }
    }

    renderError() {
        if (this.state.error) {
            return <AlertBox message={this.state.error}/>
        }
    }

    render() {
        return (
            <div>
                <h2>{T('network-config')}</h2>

                <section className="row">
                    <div>
                        <nav className="p-tabs">
                            <ul className="p-tabs__list" role="tablist">
                            {this.state.interfaces.map((iface) => {
                                let selected = "false"
                                if (this.state.selected===iface.interface) {
                                    selected = "true"
                                }
                                return (<li className="p-tabs__item" role="presentation">
                                    <a href={'#'+iface.interface} data-key={iface.interface} onClick={this.handleSelectTab} class="p-tabs__link" tabindex="0" role="tab" aria-controls="section1" aria-selected={selected}>{iface.interface}</a>
                                </li>)
                            })}
                            </ul>
                        </nav>
                    </div>
                    <div>
                        {this.renderMessage()}
                        {this.renderError()}
                        {this.state.interface.interface ?
                            <form>
                                {this.renderUse()}
                                <fieldset disabled={!this.state.interface.use}>
                                    <h3>{T('interface')} {this.state.interface.interface}</h3>
                                    <label htmlFor={"method"}>{T('method')}:</label>
                                    <select value={this.state.interface.method} onChange={this.handleMethodChange}>
                                        <option/>
                                        {methods.map((m) => {
                                            return <option value={m}>{T(m)}</option>
                                        })}
                                    </select>
                                    <label htmlFor={"nameServers"}>{T('dns')}:</label>
                                    <input name="nameServers" type="text" onChange={this.handleDNSChange}
                                           value={this.state.interface.nameServers ? this.state.interface.nameServers.toString() : ''} placeholder={T('dns-help')} />
                                    <label htmlFor={"address"}>{T('address')}:</label>
                                    <input name="address" type="text" value={this.state.interface.address} placeholder={T('address-help')} onChange={this.handleAddressChange}/>
                                    <label htmlFor={"mask"}>{T('mask')}:</label>
                                    <input name="mask" type="text" value={this.state.interface.mask} onChange={this.handleMaskChange} placeholder={T('mask-help')}/>
                                    <label for={"gateway"}>{T('gateway')}:</label>
                                    <input name="gateway" type="text" value={this.state.interface.gateway} onChange={this.handleGatewayChange}/>
                                </fieldset>
                                <button onClick={this.handleSave} className="p-button--positive">{T('save')}</button>
                            </form>
                            :
                            <h4>{T('select-interface')}</h4>
                        }
                    </div>
                    <div className="row">
                        <p className="col-10">{T('apply-desc')}</p>
                        <button onClick={this.handleApply} className="p-button--brand col-2">{T('apply')}</button>
                    </div>
                </section>
            </div>
        )
    }

}

export default Network;